crates/span/src/ast_id.rs

```diff
@@ -4,137 +4,533 @@
-register_ast_id_node! {
+register_has_name_ast_id! {
     impl AstIdNode for
-    Item, AnyHasGenericParams,
-        Adt,
-            Enum,
-                Variant,
-            Struct,
-            Union,
-        AssocItem,
-            Const,
-            Fn,
-            MacroCall,
-            TypeAlias,
-        ExternBlock,
-        ExternCrate,
-        Impl,
-        Macro,
-            MacroDef,
-            MacroRules,
-        Module,
-        Static,
-        Trait,
-        TraitAlias,
-        Use,
-    BlockExpr, ConstArg
+        Enum = name,
+        Struct = name,
+        Union = name,
+        ExternCrate = name_ref,
+        MacroDef = name,
+        MacroRules = name,
+        Module = name,
+        Static = name,
+        Trait = name,
+        TraitAlias = name
+}
+
+macro_rules! register_assoc_item_ast_id {
+    (impl AstIdNode for $($ident:ident = $name_callback:expr),+ ) => {
+        $(
+            impl AstIdNode for ast::$ident {}
+        )+
+
+        fn assoc_item_ast_id(
+            node: &SyntaxNode,
+            index_map: &mut ErasedAstIdNextIndexMap,
+            parent: Option<&ErasedFileAstId>,
+        ) -> Option<ErasedFileAstId> {
+            let kind = node.kind();
+            match_ast! {
+                match node {
+                    $(
+                        ast::$ident(node) => {
+                            let name = $name_callback(node);
+                            let name = name.as_ref().map_or("", |it| it.text_non_mutable());
+                            let properties = ErasedHasNameFileAstId {
+                                kind,
+                                name,
+                            };
+                            let result = ErasedAssocItemFileAstId {
+                                parent: parent.copied(),
+                                properties,
+                            };
+                            Some(index_map.new_id(ErasedFileAstIdKind::$ident, result))
+                        },
+                    )*
+                    _ => None,
+                }
+            }
+        }
+
+        fn should_alloc_assoc_item(node: &SyntaxNode) -> bool {
+            let kind = node.kind();
+            false $( || ast::$ident::can_cast(kind) )*
+        }
+    };
+}
+register_assoc_item_ast_id! {
+    impl AstIdNode for
+    Variant = |it: ast::Variant| it.name(),
+    Const = |it: ast::Const| it.name(),
+    Fn = |it: ast::Fn| it.name(),
+    MacroCall = |it: ast::MacroCall| it.path().and_then(|path| path.segment()?.name_ref()),
+    TypeAlias = |it: ast::TypeAlias| it.name()
 }
 
 /// Maps items' `SyntaxNode`s to `ErasedFileAstId`s and back.
 #[derive(Default)]
 pub struct AstIdMap {
-    /// Maps stable id to unstable ptr.
-    arena: Arena<SyntaxNodePtr>,
-    /// Reverse: map ptr to id.
-    map: hashbrown::HashTable<Idx<SyntaxNodePtr>>,
+    /// An arena of the ptrs and their associated ID.
+    arena: Arena<(SyntaxNodePtr, ErasedFileAstId)>,
```

Comment:
I don't think we are using the arena properties here anymore, so we can just use a vec instead here? (or event better, `Box<[]>`?

Reply:
Well we us `Idx`, and it wasn't really different previously.

Reply:
Right, well something we can consider later